### PR TITLE
ADSB: refactor (before backend split)

### PR DESCRIPTION
**Pull request by magicrub**
This is a simpler baby-step version leading up to the ADSB backend spilt (PR https://github.com/ArduPilot/ardupilot/pull/15576). This contains a variety of overdue tweaks that bring it up to speed with other libraries and will make it easier to do the split to add drivers.

### CODE REVIEW

**Comment by magicrub** [comment]
I added a few more tweaks per @peterbarker 's review of https://github.com/ArduPilot/ardupilot/pull/15576

**Comment by magicrub** [comment]
Fixed a commit msg which required a force push so I rebased while I was at it.

**Comment by magicrub** [comment]
For those of you playing along at home.. this will be the branch for the actual backend split that is on top of this PR https://github.com/magicrub/ardupilot/tree/pr/adsb_backendSplit4

**Comment by magicrub** [comment]
ping @tridge @peterbarker

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 1]
```diff
@@ -166,51 +168,73 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
+        in_state.vehicle_list = new adsb_vehicle_t[in_state.list_size_allocated];
 
         if (in_state.vehicle_list == nullptr) {
-            // dynamic RAM allocation of _vehicle_list[] failed, disable gracefully
-            hal.console->printf("Unable to initialize ADS-B vehicle list\n");
-            _enabled.set_and_notify(0);
+            // dynamic RAM allocation of in_state.vehicle_list[] failed, disable gracefully
+            hal.console->printf("Unable to initialize ADSB vehicle list\n");
+            // disable ourselves to avoid repeated allocation attempts
+            _enabled.set(0);
```
Comment (tridge) [inline libraries/AP_ADSB/AP_ADSB.cpp:199]: we no longer use a set(0) to mark a subsystem as unable to init. Instead we would check in_state.vehicle_list == nullptr in each of the API calls. This avoids users getting confused as to why their parameter is changing to zero

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 2]
```diff
@@ -166,51 +168,73 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
+
+        in_state.list_size_allocated = in_state.list_size_param;
+        in_state.vehicle_list = new adsb_vehicle_t[in_state.list_size_allocated];
 
         if (in_state.vehicle_list == nullptr) {
-            // dynamic RAM allocation of _vehicle_list[] failed, disable gracefully
-            hal.console->printf("Unable to initialize ADS-B vehicle list\n");
-            _enabled.set_and_notify(0);
+            // dynamic RAM allocation of in_state.vehicle_list[] failed, disable gracefully
+            hal.console->printf("Unable to initialize ADSB vehicle list\n");
```
Comment (tridge) [inline libraries/AP_ADSB/AP_ADSB.cpp:197]: needs to be a send_text

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 3]
```diff
@@ -36,17 +36,19 @@
...
 #define ADSB_BITBASK_RF_CAPABILITIES_1090ES_IN      (1 << 1)
 
-#if APM_BUILD_TYPE(APM_BUILD_ArduPlane)
-    #define ADSB_LIST_RADIUS_DEFAULT        10000 // in meters
-#else // APM_BUILD_TYPE(APM_BUILD_ArduCopter), Rover, Boat
-    #define ADSB_LIST_RADIUS_DEFAULT        2000 // in meters
+#ifndef ADSB_LIST_RADIUS_DEFAULT
+    #if APM_BUILD_TYPE(APM_BUILD_ArduPlane)
+        #define ADSB_LIST_RADIUS_DEFAULT        10000 // in meters
+    #else // APM_BUILD_TYPE(APM_BUILD_ArduCopter), Rover, Boat
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:49]: Nix the comment here.

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 4]
```diff
@@ -877,36 +926,47 @@ void AP_ADSB::set_callsign(const char* str, const bool append_icao)
...
 
 void AP_ADSB::handle_message(const mavlink_channel_t chan, const mavlink_message_t &msg)
 {
     switch (msg.msgid) {
-    case MAVLINK_MSG_ID_ADSB_VEHICLE:
-        handle_vehicle(msg);
+    case MAVLINK_MSG_ID_ADSB_VEHICLE: {
+        adsb_vehicle_t vehicle {};
+        mavlink_msg_adsb_vehicle_decode(&msg, &vehicle.info);
+        vehicle.last_update_ms = AP_HAL::millis() - (vehicle.info.tslc * 1000);
```
Comment (tridge) [inline libraries/AP_ADSB/AP_ADSB.cpp:949]: should be:
AP_HAL::millis() - uint32_t(vehicle.info.tslc * 1000U)

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 5]
```diff
@@ -106,19 +108,19 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
     // @Values: 0:NO_DATA,1:AppliedBySensor
     // @User: Advanced
-    AP_GROUPINFO("OFFSET_LON",   8, AP_ADSB, out_state.cfg.gpsLonOffset, UAVIONIX_ADSB_OUT_CFG_GPS_OFFSET_LON_APPLIED_BY_SENSOR),
+    AP_GROUPINFO("OFFSET_LON",   8, AP_ADSB, out_state.cfg.gpsOffsetLon, UAVIONIX_ADSB_OUT_CFG_GPS_OFFSET_LON_APPLIED_BY_SENSOR),
 
     // @Param: RF_SELECT
     // @DisplayName: Transceiver RF selection
-    // @Description: Transceiver RF selection for Rx enable and/or Tx enable. This only effects devices that can Tx and Rx. Rx-only devices override this to always be Rx-only.
-    // @Values: 0:Disabled,1:Rx-Only,2:Tx-Only,3:Rx and Tx Enabled
+    // @Description: Transceiver RF selection for Rx enable and/or Tx enable. This only effects devices that can Tx and/or Rx. Rx-only devices override this to always be Rx-only.
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:122]: The comment could do with some clarification.

It *should* be that receive-only devices ignore the Tx bit.

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 6]
```diff
@@ -166,51 +168,73 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
 {
     AP_Param::setup_object_defaults(this, var_info);
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
     if (_singleton != nullptr) {
         AP_HAL::panic("AP_ADSB must be singleton");
     }
+#endif
     _singleton = this;
+
+#ifdef ADSB_STATIC_CALLSIGN
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:178]: This looks like new code?

I do note it is in a separate commit, and its a pretty minor addition.

File: libraries/AP_ADSB/AP_ADSB.h [hunk 7]
```diff
@@ -54,20 +59,29 @@ class AP_ADSB {
...
+        return true;
     }
 
-    void set_is_auto_mode(const bool is_in_auto_mode) { out_state._is_in_auto_mode = is_in_auto_mode; }
+    bool set_max_speed(int16_t max_speed) {
+        if (out_state.cfg.was_set_externally) {
+            return false;
+        }
+        // convert m/s to knots
+        out_state.cfg.maxAircraftSpeed_knots = (float)max_speed * 1.94384f;
```
Comment (tridge) [inline libraries/AP_ADSB/AP_ADSB.h:80]: use M_PER_SEC_TO_KNOTS

File: libraries/AP_ADSB/AP_ADSB.h [hunk 8]
```diff
@@ -122,26 +132,24 @@ class AP_ADSB {
     void set_vehicle(const uint16_t index, const adsb_vehicle_t &vehicle);
 
     // Generates pseudorandom ICAO from gps time, lat, and lon
-    uint32_t genICAO(const Location &loc);
+    static uint32_t genICAO(const Location &loc);
```
Comment (tridge) [inline libraries/AP_ADSB/AP_ADSB.h:135]: make const

File: libraries/AP_ADSB/AP_ADSB.h [hunk 9]
```diff
@@ -122,26 +132,24 @@ class AP_ADSB {
...
+    bool set_callsign(const char* str, const bool append_icao);
 
     // send static and dynamic data to ADSB transceiver
     void send_configure(const mavlink_channel_t chan);
-    void send_dynamic_out(const mavlink_channel_t chan);
+    void send_dynamic_out(const mavlink_channel_t chan) const;
 
     // special helpers for uAvionix workarounds
-    uint32_t get_encoded_icao(void);
+    static uint32_t get_encoded_icao(const uint32_t icao_id);
```
Comment (tridge) [inline libraries/AP_ADSB/AP_ADSB.h:145]: make const

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 10]
```diff
@@ -166,51 +168,73 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
-    in_state.vehicle_count = 0;
     if (in_state.vehicle_list == nullptr) {
-        if (in_state.list_size_param != constrain_int16(in_state.list_size_param, 1, ADSB_VEHICLE_LIST_SIZE_MAX)) {
-            in_state.list_size_param.set_and_notify(ADSB_VEHICLE_LIST_SIZE_DEFAULT);
-            in_state.list_size_param.save();
-        }
-        in_state.list_size = in_state.list_size_param;
-        in_state.vehicle_list = new adsb_vehicle_t[in_state.list_size];
+        // sanity check param
+        in_state.list_size_param = constrain_int16(in_state.list_size_param, 1, ADSB_VEHICLE_LIST_SIZE_MAX);
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:190]: This bit is a little tricky.  We can only store signed numbers in `AP_Param`, but a list size must be non-negative.  So this constrain isn't right.  It is pre-existing, 'though, so *shrug*.

File: libraries/AP_ADSB/AP_ADSB.h [hunk 11]
```diff
@@ -152,18 +160,21 @@ class AP_ADSB {
...
         AP_Int16    list_altitude;
 
+        // index of and distance to furthest vehicle in list
+        uint16_t    furthest_vehicle_index;
+        float       furthest_vehicle_distance;
+
         // streamrate stuff
-        uint32_t    send_start_ms[MAVLINK_COMM_NUM_BUFFERS];
-        uint16_t    send_index[MAVLINK_COMM_NUM_BUFFERS];
+        uint32_t    send_start_ms[MAVLINK_COMM_NUM_BUFFERS] {};
```
Comment (tridge) [inline libraries/AP_ADSB/AP_ADSB.h:174]: wastes flash to init

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 12]
```diff
@@ -166,51 +168,73 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
+
+        in_state.list_size_allocated = in_state.list_size_param;
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:192]: Perhaps this assignment should happen after the allocation succeeds.
\

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 1 above)
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:199]: As this exists in init(), I don't think this code can get called multiple times?

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 13]
```diff
@@ -166,51 +168,73 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
     AP_Param::setup_object_defaults(this, var_info);
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
     if (_singleton != nullptr) {
         AP_HAL::panic("AP_ADSB must be singleton");
     }
+#endif
     _singleton = this;
+
+#ifdef ADSB_STATIC_CALLSIGN
+    memcpy(&out_state.cfg.callsign, ADSB_STATIC_CALLSIGN, 8);
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:179]: Should be using sizeof-the-target.

Further, what guarantees on the length of `ADSB_STATIC_CALLSIGN` are there?  Would someone defining the symbol know they exist?

**Review by peterbarker — CHANGES_REQUESTED** [review]
Got about 15% of the way through, sorry.

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:199]: I've added variable `bool init_failed` that keeps us from re-initing forever

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 6 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:178]: yes, this is new

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 10 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:190]: I don't understand. It's constraining from 1 to 100. Both are positive. note, it's not allowing 0 as a list size.

Comment (peterbarker) [reply libraries/AP_ADSB/AP_ADSB.cpp:190]: It should be `constrain_uint16`.  It doesn't really matter.

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 12 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:192]: yes, that makes sense. Doesn't actually matter since it won't be used when nullptr but good to change anyway.

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:199]: correct, but I talked to @tridge about it and it's getting a new init_failed flag so we don't actually disable the param. It will also send a GCS send_text instead this way

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp]: fixed

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 4 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp]: fixed

File: libraries/AP_ADSB/AP_ADSB.h (hunk 7 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.h]: fixed

File: libraries/AP_ADSB/AP_ADSB.h (hunk 8 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.h]: fixed

File: libraries/AP_ADSB/AP_ADSB.h (hunk 9 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.h]: fixed

File: libraries/AP_ADSB/AP_ADSB.h (hunk 11 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.h]: fixed

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 14]
```diff
@@ -957,20 +937,30 @@ bool AP_ADSB::next_sample(adsb_vehicle_t &vehicle)
...
 {
     switch (msg.msgid) {
-    case MAVLINK_MSG_ID_ADSB_VEHICLE:
-        handle_vehicle(msg);
+    case MAVLINK_MSG_ID_ADSB_VEHICLE: {
+        adsb_vehicle_t vehicle {};
+        mavlink_msg_adsb_vehicle_decode(&msg, &vehicle.info);
+        vehicle.last_update_ms = AP_HAL::millis() - (vehicle.info.tslc * 1000);
+        handle_adsb_vehicle(vehicle);
+        }
```
[the remark this answers is not in the document]
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp]: fixed

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 3 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp]: nixxed

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 5 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp]: changed

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 12 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp]: fixed, is now only assigned on success

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 13 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp]: changed to:

`#ifdef ADSB_STATIC_CALLSIGN`
    `const uint8_t callsign_len = MIN(sizeof(out_state.cfg.callsign),sizeof(ADSB_STATIC_CALLSIGN));`
    `memcpy(&out_state.cfg.callsign, ADSB_STATIC_CALLSIGN, callsign_len);`
`#endif`

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 10 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:193]: oh I see. But I'm then assigning it to an int16_t so then it's re-casting anyway.. so I think it's actually better this way.

Comment (peterbarker) [reply libraries/AP_ADSB/AP_ADSB.cpp:191]: It might *look* like an int16_t, but it is interpreted as a uint by the code.  ATM if `ADSB_VEHICLE_LIST_SIZE_MAX` was 40,000 your `constrain_int16` call would be non-sensical as the latter value would be out-of-bound for a signed 16-bit integer.  It is entirely within range of the *intent* of the parameter, an unsigned integer.  The code is rife with places where we treat AP_Int* as unsigned.

**Comment by magicrub** [comment]
rebased

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Review by magicrub — COMMENTED** [review]
back to you @peterbarker

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 15]
```diff
@@ -166,53 +172,72 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
-        if (in_state.list_size_param != constrain_int16(in_state.list_size_param, 1, ADSB_VEHICLE_LIST_SIZE_MAX)) {
-            in_state.list_size_param.set_and_notify(ADSB_VEHICLE_LIST_SIZE_DEFAULT);
-            in_state.list_size_param.save();
-        }
-        in_state.list_size = in_state.list_size_param;
-        in_state.vehicle_list = new adsb_vehicle_t[in_state.list_size];
+        // sanity check param
+        in_state.list_size_param = constrain_int16(in_state.list_size_param, 1, INT16_MAX);
+
+        in_state.vehicle_list = new adsb_vehicle_t[in_state.list_size_allocated];
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:196]: This isn't going to work.
```suggestion
        in_state.vehicle_list = new adsb_vehicle_t[in_state.list_size_param];
```

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 16]
```diff
@@ -166,53 +172,72 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
+bool AP_ADSB::check_startup()
+{
+    if (!_enabled || _init_failed) {
+        if (in_state.vehicle_list != nullptr) {
+            deinit();
+        }
+        // nothing to do
+        return false;
+    }
+    if (in_state.list_size_allocated != in_state.list_size_param) {
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:231]: This `if` should be removed, I think?

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 17]
```diff
@@ -166,53 +172,72 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
         }
+        in_state.list_size_allocated = in_state.list_size_param;
     }
-
-    furthest_vehicle_distance = 0;
-    furthest_vehicle_index = 0;
-
-    // out_state
-    set_callsign("PING1234", false);
+    in_state.furthest_vehicle_distance = 0;
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:206]: Don't need to do these initialisations if init() is only ever going to be called once.

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 18]
```diff
@@ -295,8 +298,8 @@ void AP_ADSB::update(void)
         // reset timer constantly so it never reaches 10s so it never sends
         out_state.last_config_ms = now;
 
-    } else if (out_state.cfg.ICAO_id == 0 ||
-        out_state.cfg.ICAO_id_param_prev != out_state.cfg.ICAO_id_param) {
+    } else if ((out_state.cfg.rfSelect & UAVIONIX_ADSB_OUT_RF_SELECT_TX_ENABLED) &&
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:301]: I wonder if this could wait for the next set of patches?

File: libraries/AP_Arming/AP_Arming.cpp [hunk 19]
```diff
@@ -801,6 +802,13 @@ bool AP_Arming::system_checks(bool report)
             check_failed(ARMING_CHECK_SYSTEM, report, "Scripting out of memory");
             return false;
         }
+#endif
+#if HAL_ADSB_ENABLED
+        AP_ADSB *adsb = AP::ADSB();
+        if ((adsb != nullptr) && adsb->init_failed()) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp:808]:
```suggestion
        if ((adsb == nullptr) || adsb->init_failed()) {
```

Better to be over-cautious than under-cautious.

**Review by peterbarker — CHANGES_REQUESTED** [review]
This is a nice cleanup.

**Comment by peterbarker** [comment]
```
MANUAL> arm throttle
MANUAL> APM: PreArm: ADSB out of memory
Got COMMAND_ACK: COMPONENT_ARM_DISARM: FAILED
```

(rover)

File: libraries/AP_Arming/AP_Arming.cpp (hunk 19 above)
Comment (magicrub) [reply libraries/AP_Arming/AP_Arming.cpp]: Your suggestion backfired, the nullptr check was true in rover. I changed the logic to match the other checks

**Comment by magicrub** [comment]
@peterbarker note, I rebased everything as-is but then changed this minor thing since you last review. See commits https://github.com/ArduPilot/ardupilot/pull/15610/commits/735512120510b53ac3526e4ff11bfe25b2dd9775 and https://github.com/ArduPilot/ardupilot/pull/15610/commits/cee60e8e7dc68e5d6a5f36a06ac162c46e68e217

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 20]
```diff
@@ -166,53 +172,65 @@ const AP_Param::GroupInfo AP_ADSB::var_info[] = {
...
     if (in_state.vehicle_list != nullptr) {
         delete [] in_state.vehicle_list;
         in_state.vehicle_list = nullptr;
     }
 }
 
+bool AP_ADSB::check_startup()
+{
+    if (!_enabled || _init_failed) {
+        if (in_state.vehicle_list != nullptr) {
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:222]: Minor thing - this check is redundant.

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 21]
```diff
@@ -295,8 +291,8 @@ void AP_ADSB::update(void)
         // reset timer constantly so it never reaches 10s so it never sends
         out_state.last_config_ms = now;
 
-    } else if (out_state.cfg.ICAO_id == 0 ||
-        out_state.cfg.ICAO_id_param_prev != out_state.cfg.ICAO_id_param) {
+    } else if ((out_state.cfg.rfSelect & UAVIONIX_ADSB_OUT_RF_SELECT_TX_ENABLED) &&
+            (out_state.cfg.ICAO_id == 0 || out_state.cfg.ICAO_id_param_prev != out_state.cfg.ICAO_id_param)) {
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:295]: Indenting is a little stuffed here.

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 22]
```diff
@@ -434,6 +441,8 @@ void AP_ADSB::handle_adsb_vehicle(const adsb_vehicle_t &vehicle)
     const uint16_t required_flags_position = ADSB_FLAGS_VALID_COORDS | ADSB_FLAGS_VALID_ALTITUDE;
     const bool detected_ourself = (out_state.cfg.ICAO_id != 0) && ((uint32_t)out_state.cfg.ICAO_id == vehicle.info.ICAO_address);
 
+    vehicle.last_update_ms = now - uint32_t(vehicle.info.tslc * 1000U);
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:444]: This is worth of an ~~apology~~ comment.  You're modifying a parameter-taken-by-reference.  You're copying it later, but the thing passed in is still modified...

Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:444]: tslc is coming from external sources.  It should probably be validated before being used here...  this is a pre-existing issue so not a blocker.

File: libraries/AP_ADSB/AP_ADSB.cpp [hunk 23]
```diff
@@ -496,36 +505,82 @@ void AP_ADSB::handle_adsb_vehicle(const adsb_vehicle_t &vehicle)
...
+                ADSB_FLAGS_VERTICAL_VELOCITY_VALID |
+                ADSB_FLAGS_BARO_VALID;
+                // all flags set except ADSB_FLAGS_SOURCE_UAT and ADSB_FLAGS_SIMULATED
+    }
+
+    in_state.vehicle_list[index].last_update_ms = vehicle.last_update_ms;
+
+    // use |= so we allow partial sets and those don't override previous valid settings
+    mavlink_adsb_vehicle_t &adsb_vehicle = in_state.vehicle_list[index].info;
+    adsb_vehicle.flags |= flags;
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.cpp:538]: Doesn't this mean that if a flag gets turned off by the external vehicle (e.g. they've lost their position) we won't notice that?

File: libraries/AP_ADSB/AP_ADSB.h [hunk 24]
```diff
@@ -96,20 +117,16 @@ class AP_ADSB {
...
-    // get singleton instance
-    static AP_ADSB *get_singleton(void) {
-        return _singleton;
-    }
-
 private:
     static AP_ADSB *_singleton;
 
-    // initialize _vehicle_list
+    // initialize/reinitialize vehicle_list
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.h:123]: Comment now incorrect.

File: libraries/AP_ADSB/AP_ADSB.h [hunk 25]
```diff
@@ -172,18 +191,18 @@ class AP_ADSB {
...
-        bool        _is_in_auto_mode;
+        bool        is_in_auto_mode;
 
         // ADSB-OUT configuration
         struct {
             int32_t     ICAO_id;
             AP_Int32    ICAO_id_param;
             int32_t     ICAO_id_param_prev = -1; // assume we never send
-            char        callsign[MAVLINK_MSG_UAVIONIX_ADSB_OUT_CFG_FIELD_CALLSIGN_LEN]; //Vehicle identifier (8 characters, null terminated, valid characters are A-Z, 0-9, " " only).
+            char        callsign[MAVLINK_MSG_UAVIONIX_ADSB_OUT_CFG_FIELD_CALLSIGN_LEN] {}; //Vehicle identifier (8 characters, null terminated, valid characters are A-Z, 0-9, " " only).
```
Comment (peterbarker) [inline libraries/AP_ADSB/AP_ADSB.h:201]: Does this actually need zeroing?

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 22 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:444]: I'll just remove the brand new commits for this and keep the old behavior.

Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:444]: I'll just remove the brand new commits for this and keep the old behavior.

File: libraries/AP_ADSB/AP_ADSB.h (hunk 24 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.h:123]: fixed

File: libraries/AP_ADSB/AP_ADSB.h (hunk 25 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.h:201]: nope. Removed.

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 23 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:538]: yes, and then the uAvionix will send us zeros for lat/lng (for better or worse). But I agree we should catch that case on our end just in case

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 20 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:222]: sure, I'll remove it. A part of me wants to leave it for clarity but I agree it's redundant.

Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:222]: actually.. I have a feeling this will backfire in the future. It's ripe with human-error. I'd rather not call deinit() constantly in case something gets added to deinit in the future.

**Comment by magicrub** [comment]
addressed suggestions from @peterbarker and squashed and rebased.

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 23 above)
Comment (peterbarker) [reply libraries/AP_ADSB/AP_ADSB.cpp:536]: Are you going to patch that, then?

If we are passing these flags onto a GCS, haven't we lost the information that the original vehicle is no longer (e.g.) giving a position estimate?

**Review by peterbarker — COMMENTED** [review]
I've put this on a quadcopter with a ping sensor on it, and its definitely not fundamentally broken - I'm still getting vehicles to the GCS.

File: libraries/AP_ADSB/AP_ADSB.cpp (hunk 23 above)
Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:536]: Yeah. Gimme a couple hours.

Comment (magicrub) [reply libraries/AP_ADSB/AP_ADSB.cpp:536]: Yes. Although I'm tempted to undo this just to get this damn thing in.

**Comment by peterbarker** [comment]
@magicrub did you test this with a static callsign?  Did the static callsign appear on a second ADSB receiver?

**Comment by magicrub** [comment]
I have been using a static callsign on my fork, where this originally came from, but have not tested it in the last 2 weeks beyond compile checks.

**Comment by peterbarker** [comment]
> I have been using a static callsign on my fork, where this originally came from, but have not tested it in the last 2 weeks beyond compile checks.

Did the static callsign turn up on an ADSB receiver?

How much work is it to test it again with this branch?

**Comment by magicrub** [comment]
I can retest it. Anything else you'd like tested?

**Comment by peterbarker** [comment]
Merged, thanks!

**Comment by magicrub** [comment]
Thanks!
